fud/sim/testbench.cpp

```diff
@@ -1,84 +1,98 @@
 #include "Vmain.h"
 #include "verilated.h"
 #include "verilated_vcd_c.h"
-#include <stdio.h>
+#include <algorithm>
+#include <array>
+#include <cstdint>
+#include <iostream>
 #include <memory>
 
-// Keep track of time: https://www.veripool.org/wiki/verilator/Manual-verilator#CONNECTING-TO-C
-vluint64_t main_time = 0;
-double sc_time_stamp() { return main_time; }
+// Keep track of time:
+// https://www.veripool.org/wiki/verilator/Manual-verilator#CONNECTING-TO-C
+vluint64_t GLOBAL_sc_time = 0;
+// Accessed by the underlying test bench.
+double sc_time_stamp() { return GLOBAL_sc_time; }
 
-int main(int argc, char **argv, char **env) {
-  int i = 0;
-  int clk;
-  Verilated::commandArgs(argc, argv);
-
-  // init top verilog instance
-  std::unique_ptr<Vmain> top = std::make_unique<Vmain>();
+// This shows simulation progress by printing the number of cycles at certain
+// intervals.
+void cycle_tracker(const uint64_t current_cycle, const uint64_t n_cycles) {
+  auto print_simulation_at = [&](int64_t cycle) {
+    std::cout << "[Verilator] In-progress: Simulated " << cycle << " cycles\n";
```

Comment:
How often does this print? Isn't this going to clutter the log if it prints every cycle?

Reply:
For 5e8, it probably won't print at all for anything in our test suite. It prints at the intervals listed below, e.g. `5e8 / 4`, ... 

This is mostly for the bigger cases such as lenet to ensure something didn't time out

Reply:
Got it. LGTM for everything else.